File: web/src/app/core/srt-parser.ts
// The wire format the model is sent and the validation of what comes back:
// parsing and serializing SRT, the timestamp-free "lite" batch format, batch
// splitting, and the checks a reply must pass before it is accepted.

export interface SubtitleBlock {
  number: number;
  timestamp: string;
  text: string;
}

export interface ValidationResult {
  ok: boolean;
  error: string;
}

const TIMESTAMP_RE = /^\d{2}:\d{2}:\d{2},\d{3}\s*-->\s*\d{2}:\d{2}:\d{2},\d{3}$/;
// Unanchored: catches a timestamp anywhere inside a returned block's text.
const TIMESTAMP_IN_TEXT_RE = /\d{2}:\d{2}:\d{2},\d{3}\s*-->\s*\d{2}:\d{2}:\d{2},\d{3}/;
const BLOCK_NUMBER_RE = /^[+-]?\d+$/;

/** Cue text as the rest of the pipeline expects it: LF breaks, no trailing
 * blank lines. Interior blanks stay — serializeLite collapses those. Every
 * format parser runs its cue text through this. */
export function normalizeCueText(text: string): string {
  const lines = text.replace(/\r\n?/g, '\n').split('\n');
  while (lines.length && !lines[lines.length - 1]!.trim()) lines.pop();
  return lines.join('\n');
}

export function parseSrt(content: string): SubtitleBlock[] {
  content = normalizeCueText(content);
  if (content.charCodeAt(0) === 0xfeff) {
    content = content.slice(1);
  }

  const rawBlocks = content.trim().split(/\n\n+/);
  const blocks: SubtitleBlock[] = [];

  for (const raw of rawBlocks) {
    const lines = raw.trim().split('\n');
    if (lines.length < 2) continue;

    const number = parseInt(lines[0].trim(), 10);
    if (isNaN(number)) continue;

    const timestamp = lines[1].trim();
    if (!TIMESTAMP_RE.test(timestamp)) continue;

    const text = lines.slice(2).join('\n');

    blocks.push({ number, timestamp, text });
  }

  return blocks;
}

export function serializeSrt(blocks: SubtitleBlock[]): string {
  return (
    blocks.map((b) => `${b.number}\n${b.timestamp}\n${b.text}`).join('\n\n') +
    '\n'
  );
}

// Wire format: number + text only. Timestamps are stripped before sending
// because small models sometimes corrupt them; callers reattach positionally.
// A blank line inside a cue is dropped: it would forge a block boundary.
export function serializeLite(blocks: SubtitleBlock[]): string {
  return blocks
    .map((b) => {
      const text = b.text.split('\n').filter((line) => line.trim()).join('\n');
      return `${b.number}\n${text}`;
    })
    .join('\n\n') + '\n';
}

export function parseLite(content: string): SubtitleBlock[] {
  content = normalizeCueText(content);
  if (content.charCodeAt(0) === 0xfeff) content = content.slice(1);

  const rawBlocks = content.trim().split(/\n\n+/);
  const blocks: SubtitleBlock[] = [];

  for (const raw of rawBlocks) {
    const lines = raw.trim().split('\n');
    if (lines.length < 1) continue;

    // Strict: "1) text" would otherwise parse as an empty block that validates.
    const numberLine = lines[0].trim();
    if (!BLOCK_NUMBER_RE.test(numberLine)) continue;
    const number = parseInt(numberLine, 10);

    const text = lines.slice(1).join('\n');
    blocks.push({ number, timestamp: '', text });
  }
  return blocks;
}

export function splitBatches(
  blocks: SubtitleBlock[],
  batchSize = 15
): SubtitleBlock[][] {
  const batches: SubtitleBlock[][] = [];
  for (let i = 0; i < blocks.length; i += batchSize) {
    batches.push(blocks.slice(i, i + batchSize));
  }
  return batches;
}

export function validateBatch(
  inputBlocks: SubtitleBlock[],
  outputBlocks: SubtitleBlock[]
): ValidationResult {
  if (inputBlocks.length !== outputBlocks.length) {
    return {
      ok: false,
      error: `Block count mismatch: expected ${inputBlocks.length}, got ${outputBlocks.length}`,
    };
  }

  for (let i = 0; i < inputBlocks.length; i++) {
    if (inputBlocks[i].number !== outputBlocks[i].number) {
      return {
        ok: false,
        error: `Block number mismatch at index ${i}: expected ${inputBlocks[i].number}, got ${outputBlocks[i].number}`,
      };
    }
  }

  // Wire-parsed blocks have no timestamp yet; compare only once one is set.
  for (let i = 0; i < inputBlocks.length; i++) {
    if (outputBlocks[i].timestamp && inputBlocks[i].timestamp !== outputBlocks[i].timestamp) {
      return {
        ok: false,
        error:
          `Timestamp modified at block ${inputBlocks[i].number}: ` +
          `expected '${inputBlocks[i].timestamp}', got '${outputBlocks[i].timestamp}'`,
      };
    }
  }

  // The model can shift blocks and leave a tail blank with count/numbers intact.
  for (let i = 0; i < inputBlocks.length; i++) {
    if (inputBlocks[i].text.trim() && !outputBlocks[i].text.trim()) {
      return {
        ok: false,
        error: `Empty output at block ${inputBlocks[i].number} (input was non-empty)`,
      };
    }
  }

  // Wire format has no timestamps, so one in the text is an invented line.
  for (let i = 0; i < inputBlocks.length; i++) {
    if (TIMESTAMP_IN_TEXT_RE.test(outputBlocks[i].text)) {
      return {
        ok: false,
        error: `Timestamp line leaked into text at block ${inputBlocks[i].number}`,
      };
    }
  }

  return { ok: true, error: '' };
}

File: cli/core/srt_parser.py
"""The SRT wire format: parsing, serializing, batch splitting, and the
validation of a model's reply against the batch it was sent."""

from __future__ import annotations

import re
from dataclasses import dataclass

TIMESTAMP_RE = re.compile(
    r"^\d{2}:\d{2}:\d{2},\d{3}\s*-->\s*\d{2}:\d{2}:\d{2},\d{3}$"
)
# Unanchored: catches a timestamp the model echoed anywhere inside a block.
TIMESTAMP_IN_TEXT_RE = re.compile(
    r"\d{2}:\d{2}:\d{2},\d{3}\s*-->\s*\d{2}:\d{2}:\d{2},\d{3}"
)


@dataclass
class SubtitleBlock:
    number: int
    timestamp: str
    text: str


def normalize_text(text: str) -> str:
    """One newline convention, no BOM, no trailing blank lines. Every parser
    runs its cue text through this: a stray CR reads as a second line, so it
    doubles the cue's line count and pushes a closing tag off the end."""
    if text.startswith("\ufeff"):
        text = text[1:]
    lines = text.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    while lines and not lines[-1].strip():
        lines.pop()
    return "\n".join(lines)


def parse_srt(content: str) -> list[SubtitleBlock]:
    raw_blocks = re.split(r"\n\n+", normalize_text(content).strip())

    blocks: list[SubtitleBlock] = []
    for raw in raw_blocks:
        lines = raw.strip().split("\n")
        if len(lines) < 2:
            continue
        try:
            number = int(lines[0].strip())
        except ValueError:
            continue
        timestamp = lines[1].strip()
        if not TIMESTAMP_RE.match(timestamp):
            continue
        text = "\n".join(lines[2:]) if len(lines) > 2 else ""
        blocks.append(SubtitleBlock(number=number, timestamp=timestamp, text=text))

    return blocks


def serialize_srt(blocks: list[SubtitleBlock]) -> str:
    parts: list[str] = []
    for block in blocks:
        parts.append(f"{block.number}\n{block.timestamp}\n{block.text}")
    return "\n\n".join(parts) + "\n"


def _collapse_blank_lines(text: str) -> str:
    """A blank line inside a block would forge a block boundary in the wire format."""
    return "\n".join(line for line in text.split("\n") if line.strip())


# Wire format: number + text only. Timestamps are stripped before sending
# because small models sometimes corrupt them; callers reattach positionally.
def serialize_lite(blocks: list[SubtitleBlock]) -> str:
    return "\n\n".join(
        f"{b.number}\n{_collapse_blank_lines(b.text)}" for b in blocks
    ) + "\n"


def parse_lite(content: str) -> list[SubtitleBlock]:
    raw_blocks = re.split(r"\n\n+", normalize_text(content).strip())

    blocks: list[SubtitleBlock] = []
    for raw in raw_blocks:
        lines = raw.strip().split("\n")
        if not lines:
            continue
        try:
            number = int(lines[0].strip())
        except ValueError:
            continue
        text = "\n".join(lines[1:]) if len(lines) > 1 else ""
        blocks.append(SubtitleBlock(number=number, timestamp="", text=text))
    return blocks


def split_batches(blocks: list[SubtitleBlock], batch_size: int = 15) -> list[list[SubtitleBlock]]:
    return [blocks[i : i + batch_size] for i in range(0, len(blocks), batch_size)]


@dataclass
class ValidationResult:
    ok: bool
    error: str = ""


def validate_batch(
    input_blocks: list[SubtitleBlock],
    output_blocks: list[SubtitleBlock],
) -> ValidationResult:
    if len(input_blocks) != len(output_blocks):
        return ValidationResult(
            ok=False,
            error=f"Block count mismatch: expected {len(input_blocks)}, got {len(output_blocks)}",
        )

    for i, (inp, out) in enumerate(zip(input_blocks, output_blocks, strict=True)):
        if inp.number != out.number:
            return ValidationResult(
                ok=False,
                error=(f"Block number mismatch at index {i}: "
                       f"expected {inp.number}, got {out.number}"),
            )

    # Wire-parsed blocks carry no timestamp; callers reattach after validation passes.
    for inp, out in zip(input_blocks, output_blocks, strict=True):
        if out.timestamp and inp.timestamp != out.timestamp:
            return ValidationResult(
                ok=False,
                error=(f"Timestamp modified at block {inp.number}: "
                       f"expected '{inp.timestamp}', got '{out.timestamp}'"),
            )

    # Catches silent data-loss where the model shifts blocks and leaves a
    # tail block blank while preserving count/numbers/timestamps.
    for inp, out in zip(input_blocks, output_blocks, strict=True):
        if inp.text.strip() and not out.text.strip():
            return ValidationResult(
                ok=False,
                error=f"Empty output at block {inp.number} (input was non-empty)",
            )

    # The wire format has no timestamp lines, so one in the text is model leakage.
    for inp, out in zip(input_blocks, output_blocks, strict=True):
        if TIMESTAMP_IN_TEXT_RE.search(out.text):
            return ValidationResult(
                ok=False,
                error=f"Timestamp line leaked into text at block {inp.number}",
            )

    return ValidationResult(ok=True)
